### PR TITLE
pkg/trace/sampler: catch rare non p1 trace chunks

### PR DESCRIPTION
**Pull request by raphaelgavache**
### What does this PR do?

Add a sampler that will catch trace chunks with rare measured/top_level spans that are not caught by the priority sampler.

### Motivation

What inspired you to submit this pull request?

### Additional Notes

Anything else we should know when reviewing?

### CODE REVIEW

**Review by gbbr — COMMENTED** [review]
Just some initial thoughts...

File: pkg/trace/sampler/exception_sampler.go [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+	mu      sync.RWMutex
+	limiter *rate.Limiter
+	seen    map[Signature]*seenSpans
+
+	tickStats *time.Ticker
+	hits      int64
+	misses    int64
+}
+
+// NewExceptionSampler returns a NewExceptionSampler that ensures that we samplexs combinations
```
Comment (jdgumz) [inline pkg/trace/sampler/exception_sampler.go:44]: nit: typo in `sample`?

File: pkg/trace/sampler/exception_sampler.go [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+	// exceptionSamplerTPS traces per second allowed by the rate limiter.
+	exceptionSamplerTPS = 5
+	// exceptionSamplerBurst sizes the token store used by the rate limiter.
+	exceptionSamplerBurst = 50
+	exceptionKey          = "_dd.exception"
+)
+
+// ExceptionSampler samples traces that are not caught by the Priority sampler.
+// It ensures that we sample traces for each combination of
+// (env, service, name, resource, error type, http status) seen on a measured span.
```
Comment (jdgumz) [inline pkg/trace/sampler/exception_sampler.go:32]: nit: `top-level or measured`?

File: pkg/trace/sampler/exception_sampler_test.go [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+	}
+	testTime := time.Unix(13829192398, 0)
+	testCases := []testCase{
+		{"blocked-p1", false, testTime, map[string]float64{KeySamplingPriority: 1, "_top_level": 1}},
+		{"p0-blocked-by-p1", false, testTime, map[string]float64{"_top_level": 1}},
+		{"p1-ttl-before-expiration", false, testTime.Add(priorityTTL), map[string]float64{"_top_level": 1}},
+		{"p1-ttl-expired", true, testTime.Add(priorityTTL + time.Nanosecond), map[string]float64{"_top_level": 1}},
+		{"p0-ttl-active", false, testTime.Add(priorityTTL + time.Nanosecond), map[string]float64{"_top_level": 1}},
+		{"p0-ttl-before-expiration", false, testTime.Add(priorityTTL + defaultTTL + time.Nanosecond), map[string]float64{"_top_level": 1}},
+		{"p0-ttl-expired", true, testTime.Add(priorityTTL + defaultTTL + 2*time.Nanosecond), map[string]float64{"_top_level": 1}},
```
Comment (jdgumz) [inline pkg/trace/sampler/exception_sampler_test.go:27]: Do we want to add some measured spans here as well?

File: pkg/trace/sampler/exception_sampler.go [hunk 4]
```diff
@@ -0,0 +1,204 @@
...
+	// ttlRenewalPeriod prevents from continuously updating expire times for each span seen.
+	ttlRenewalPeriod = 1 * time.Minute
+	// exceptionSamplerTPS traces per second allowed by the rate limiter.
+	exceptionSamplerTPS = 5
+	// exceptionSamplerBurst sizes the token store used by the rate limiter.
+	exceptionSamplerBurst = 50
+	exceptionKey          = "_dd.exception"
+)
+
+// ExceptionSampler samples traces that are not caught by the Priority sampler.
```
Comment (jdgumz) [inline pkg/trace/sampler/exception_sampler.go:30]: Why is that the case?

File: pkg/trace/sampler/exception_sampler.go [hunk 5]
```diff
@@ -0,0 +1,204 @@
...
+		ss.shrink()
+	}
+	ss.mu.Unlock()
+}
+
+// shrink limits the cardinality of signatures considered and the memory usage.
+// This ensure that a service with high cardinality of resources does not consume
+// all sampling tokens. The cardinality limit matches a backend limit.
+// This function is not thread safe and should be called between locks
+func (ss *seenSpans) shrink() {
```
Comment (jdgumz) [inline pkg/trace/sampler/exception_sampler.go:182]: Do we have tests for this function? Or if it's part of a larger functional test, do we know it's shrinking properly?

**Review by jdgumz — COMMENTED** [review]
Looks fine overall! :+1: on considering both top-level and measured spans.

File: pkg/trace/sampler/exception_sampler.go (hunk 5 above)
Comment (furmmon) [reply pkg/trace/sampler/exception_sampler.go:183]: Yes it's the TestCardinalityLimit test

File: pkg/trace/sampler/exception_sampler.go [hunk 6]
```diff
@@ -0,0 +1,205 @@
...
+	s = &seenSpans{expires: make(map[spanHash]time.Time)}
+	e.mu.Lock()
+	e.seen[shardSig] = s
+	e.mu.Unlock()
+	return s
+}
+
+func (e *ExceptionSampler) report() {
+	metrics.Count("datadog.trace_agent.sampler.exception.hits", atomic.SwapInt64(&e.hits, 0), nil, 1)
+	metrics.Count("datadog.trace_agent.sampler.exception.misses", atomic.SwapInt64(&e.misses, 0), nil, 1)
```
Comment (bmermet) [inline pkg/trace/sampler/exception_sampler.go:149]: What do you think of adding a metric or a log when shrink is enabled to ease up troubleshooting?

**Review by bmermet — DISMISSED** [review]

**Review by bmermet — DISMISSED** [review]

File: pkg/trace/sampler/exception_sampler.go [hunk 7]
```diff
@@ -0,0 +1,210 @@
...
+	"github.com/DataDog/datadog-agent/pkg/trace/traceutil"
+	"golang.org/x/time/rate"
+)
+
+const (
+	// cardinalityLimit limits the number of spans considered per combination of (env, service).
+	cardinalityLimit = 1000
+	// defaultTTL limits the frequency at which we sample a same span (env, service, name, rsc, ...).
+	defaultTTL = 2 * time.Minute
+	// priorityTTL allows to blacklist p1 spans that are sampled entirely.
```
Comment (gbbr) [inline pkg/trace/sampler/exception_sampler.go:19]:
```suggestion
	// priorityTTL allows to blacklist p1 spans that are sampled entirely, for this period.
```
Is this right?

File: pkg/trace/sampler/exception_sampler.go [hunk 8]
```diff
@@ -0,0 +1,210 @@
...
+	priorityTTL = 10 * time.Minute
+	// ttlRenewalPeriod prevents from continuously updating expire times for each span seen.
```
Comment (gbbr) [inline pkg/trace/sampler/exception_sampler.go:21]:
```suggestion
	// ttlRenewalPeriod specifies the frequency at which we will upload cached entries.
```
I feel the current documentation has no relation to the time value.

File: pkg/trace/sampler/exception_sampler.go [hunk 9]
```diff
@@ -0,0 +1,210 @@
...
+	return e.handleTrace(now, env, t)
+}
+
+// Start starts reporting stats
+func (e *ExceptionSampler) Start() {
+	go func() {
+		for range e.tickStats.C {
+			e.report()
+		}
+	}()
```
Comment (gbbr) [inline pkg/trace/sampler/exception_sampler.go:75]: Can we move this code into `NewExceptionSampler` and have only `Stop` ? It feels weird to have to call `New` all the time followed by `Start`. If that complicates your test code or prolongs things unnecessarily, it's fine to leave it for now.

**Review by gbbr — DISMISSED** [review]
Really nice and clean Raphael ❤️Easy to read and understand.

**Comment by furmmon** [comment]
I rebased on master as the score sampler was reintroduced for traces without priority.
This is the important commit to review carefully as it changes the sampling logic
https://github.com/DataDog/datadog-agent/pull/5234/commits/ed542a638c34b700a28eb9d169ae6404344c73e0

**Comment by gbbr** [comment]
Latest commit LGTM to me.

**Review by bmermet — DISMISSED** [review]

**Review by gbbr — COMMENTED** [review]
@furmmon please make sure you address the typos and things identified by @jdgumz

File: pkg/trace/agent/agent.go [hunk 10]
```diff
@@ -262,21 +265,35 @@ func (a *Agent) sample(ts *info.TagStats, pt ProcessedTrace) {
...
+func (a *Agent) runSamplers(pt ProcessedTrace) (bool, float64) {
 	if _, ok := pt.GetSamplingPriority(); ok {
-		sampledPriority, ratePriority = a.PrioritySampler.Add(pt)
+		return a.samplePriorityTrace(pt)
 	}
+	return a.sampleNoPriorityTrace(pt)
+}
 
+// samplePriorityTrace samples traces with priority set on them. PrioritySampler and
+// ErrorSampler are run in parallel. The ExceptionSampler catches traces with rate top-level
```
Comment (LotharSee) [inline pkg/trace/agent/agent.go:276]: Nit: s/rate/rare.

File: pkg/trace/sampler/exception_sampler.go [hunk 11]
```diff
@@ -0,0 +1,207 @@
...
+	// exceptionSamplerBurst sizes the token store used by the rate limiter.
+	exceptionSamplerBurst = 50
+	exceptionKey          = "_dd.exception"
+)
+
+// ExceptionSampler samples traces that are not caught by the Priority sampler.
+// It ensures that we sample traces for each combination of
+// (env, service, name, resource, error type, http status) seen on a top level or measured span
+// for which we did not see any span with a priority > 0 (sampled by Priority).
+// The resulting sampled traces will likely be incomplete.
```
Comment (LotharSee) [inline pkg/trace/sampler/exception_sampler.go:34]: Nit: Worth adding as a comment here that these are flagged with `exceptionKey` as a span metric.

File: pkg/trace/sampler/exception_sampler.go [hunk 12]
```diff
@@ -0,0 +1,207 @@
...
+	}()
+	return e
+}
+
+// Add samples a trace and returns true if trace was sampled (should be kept)
+func (e *ExceptionSampler) Add(env string, root *pb.Span, t pb.Trace) (sampled bool) {
+	return e.add(time.Now(), env, root, t)
+}
+
+func (e *ExceptionSampler) add(now time.Time, env string, root *pb.Span, t pb.Trace) (sampled bool) {
```
Comment (LotharSee) [inline pkg/trace/sampler/exception_sampler.go:67]: When priority == -1, don't we want to drop it fully?
That case seems to go to `handleTrace`.

File: pkg/trace/sampler/exception_sampler.go [hunk 13]
```diff
@@ -0,0 +1,207 @@
...
+		if !sampled {
+			sampled = e.sampleSpan(now, env, s)
+			continue
+		}
+		e.addSpan(expire, env, s)
+	}
+	return sampled
+}
+
+func (e *ExceptionSampler) addSpan(expire time.Time, env string, s *pb.Span) {
```
Comment (LotharSee) [inline pkg/trace/sampler/exception_sampler.go:106]: Nit: some extra comments around all these internal methods (add / addSpan/ handle(Priority)Trace could help understand the logic.
At the end of the day, it's about keeping track of all past span signatures and keeping the trace only if it has been a long time we haven't seen it. Here I needed to read the whole file to get to understand it.

File: pkg/trace/sampler/exception_sampler.go (hunk 12 above)
Comment (furmmon) [reply pkg/trace/sampler/exception_sampler.go:67]: negative priorities are caught before on the agent loop https://github.com/DataDog/datadog-agent/blob/master/pkg/trace/agent/agent.go#L230-L232
They get dropped for events and traces

File: pkg/trace/sampler/exception_sampler.go [hunk 14]
```diff
@@ -0,0 +1,207 @@
...
+
+	"github.com/DataDog/datadog-agent/pkg/trace/metrics"
+	"github.com/DataDog/datadog-agent/pkg/trace/pb"
+	"github.com/DataDog/datadog-agent/pkg/trace/traceutil"
+	"golang.org/x/time/rate"
+)
+
+const (
+	// cardinalityLimit limits the number of spans considered per combination of (env, service).
+	cardinalityLimit = 1000
```
Comment (LotharSee) [inline pkg/trace/sampler/exception_sampler.go:16]: How did we pick that value? What's the estimated extra memory usage? Should it be more/less?

Comment (LotharSee) [reply pkg/trace/sampler/exception_sampler.go:16]: Figured it out looking at the shrink method.

File: pkg/trace/sampler/exception_sampler.go [hunk 15]
```diff
@@ -0,0 +1,207 @@
...
+		return s
+	}
+	s = &seenSpans{expires: make(map[spanHash]time.Time), totalSamplerShrinks: &e.shrinks}
+	e.mu.Lock()
+	e.seen[shardSig] = s
+	e.mu.Unlock()
+	return s
+}
+
+func (e *ExceptionSampler) report() {
```
Comment (LotharSee) [inline pkg/trace/sampler/exception_sampler.go:145]: How helpful are these metrics? Do we consider they have value in the production environment of any user?

Comment (LotharSee) [inline pkg/trace/sampler/exception_sampler.go:145]: Also, do we have somewhere the metric of how many traces the ExceptionSampler would have picked up?

Comment (furmmon) [reply pkg/trace/sampler/exception_sampler.go:145]: Yes we have the receiver total of P0 traces received

Comment (furmmon) [reply pkg/trace/sampler/exception_sampler.go:145]: Those metrics will be internal and are meant to help us debug missing traces (resource pages without traces)

**Review by LotharSee — APPROVED** [review]

**Pull request merged**